include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,297 @@
+    {
+        GKO_ASSERT_BATCH_HAS_SQUARE_DIMENSIONS(system_matrix_);
+
+        using value_type = typename ConcreteSolver::value_type;
+        using Identity = matrix::Identity<value_type>;
+        using real_type = remove_complex<value_type>;
+
+        if (common_params.generated_prec) {
+            GKO_ASSERT_BATCH_EQUAL_DIMENSIONS(common_params.generated_prec,
+                                              this);
+            preconditioner_ = std::move(common_params.generated_prec);
+        } else if (common_params.prec_factory) {
+            preconditioner_ =
+                common_params.prec_factory->generate(system_matrix_);
+        } else {
+            auto id = Identity::create(exec, system_matrix->get_size());
+            preconditioner_ = std::move(id);
+        }
+    }
+
+    void apply_impl(const MultiVector<ValueType>* b,
+                    MultiVector<ValueType>* x) const
+    {
+        using value_type = ValueType;
+        using Vector = MultiVector<value_type>;
+        using res_log_type = double;
+
+        auto exec = this->get_executor();
+
+        const size_type num_rhs = b->get_common_size()[0];
+        const size_type num_batch_items = b->get_num_batch_items();
+        batch_dim<2> batch_size(num_batch_items, dim<2>{1, num_rhs});
+
+        BatchInfo info;
+        info.logdata =
+            std::move(std::make_unique<log::BatchLogData<res_log_type>>());
+        auto concrete_logdata =
+            static_cast<log::BatchLogData<res_log_type>*>(info.logdata.get());
+        concrete_logdata->res_norms =
+            MultiVector<res_log_type>::create(this->get_executor(), batch_size);
+        concrete_logdata->iter_counts.set_executor(this->get_executor());
+        concrete_logdata->iter_counts.resize_and_reset(num_rhs *
+                                                       num_batch_items);
+
+        this->solver_apply(b, x, &info);
+
+        this->template log<gko::log::Logger::batch_solver_completed>(
+            concrete_logdata->iter_counts, concrete_logdata->res_norms.get());
+    }
+
+    void apply_impl(const MultiVector<ValueType>* alpha,
+                    const MultiVector<ValueType>* b,
+                    const MultiVector<ValueType>* beta,
+                    MultiVector<ValueType>* x) const
+    {
+        auto x_clone = x->clone();
+        this->apply(b, x_clone.get());
+        x->scale(beta);
+        x->add_scaled(alpha, x_clone.get());
+    }
+
+    /**
+     * Calls the concrete solver on the given system (not necessarily on
+     * system_matrix_).
+     *
+     * @param mtx  Left-hand side matrix for the linear solve.
+     * @param b  Right-hand side vector.
+     * @param x  Solution vector and initial guess.
+     * @param info  Batch logging information.
+     */
+    virtual void solver_apply(const MultiVector<ValueType>* b,
+                              MultiVector<ValueType>* x,
+                              BatchInfo* const info) const = 0;
```

Comment:
Either remove the const or put it in front of `BatchInfo *`, if that is what you wanted to do.